File: backend-api/internal/models/product.go
package models

// ProductVariant represents a product variation (size/color combination)
type ProductVariant struct {
	ID            string   `json:"id" firestore:"id"`
	Color         string   `json:"color" firestore:"color"`
	Size          string   `json:"size" firestore:"size"`
	Price         float64  `json:"price" firestore:"price"`
	SalePrice     float64  `json:"sale_price,omitempty" firestore:"sale_price,omitempty"`
	SKU           string   `json:"sku" firestore:"sku"`
	StockQuantity int      `json:"stock_quantity" firestore:"stock_quantity"`
	Images        []string `json:"images,omitempty" firestore:"images,omitempty"`
	Available     bool     `json:"available" firestore:"available"`
}

type Product struct {
	ID               string                 `json:"id" firestore:"-"`
	SKU              string                 `json:"sku" firestore:"sku"`
	Name             string                 `json:"name" firestore:"name"`
	Slug             string                 `json:"slug" firestore:"slug"`
	Description      string                 `json:"description" firestore:"description"`
	ShortDescription string                 `json:"short_description" firestore:"short_description"`
	Price            float64                `json:"price" firestore:"price"`
	SalePrice        interface{}            `json:"sale_price" firestore:"sale_price"`
	ManageStock      bool                   `json:"manage_stock" firestore:"manage_stock"`
	StockQuantity    int                    `json:"stock_quantity" firestore:"stock_quantity"`
	StockStatus      string                 `json:"stock_status" firestore:"stock_status"`
	Featured         bool                   `json:"featured" firestore:"featured"`
	Status           string                 `json:"status" firestore:"status"`
	Images           []string               `json:"images" firestore:"images"`
	Categories       []string               `json:"categories" firestore:"categories"`
	Subcategories    []string               `json:"subcategories" firestore:"subcategories"`
	Tags             []string               `json:"tags" firestore:"tags"`
	Attributes       []map[string]interface{} `json:"attributes" firestore:"attributes"`
	Dimensions       map[string]interface{}   `json:"dimensions" firestore:"dimensions"`
	Weight           map[string]interface{}   `json:"weight" firestore:"weight"`
	CreatedAt        interface{}            `json:"created_at" firestore:"created_at"`
	UpdatedAt        interface{}            `json:"updated_at" firestore:"updated_at"`
	ParsedDescription map[string]interface{} `json:"parsed_description,omitempty" firestore:"parsed_description,omitempty"`
	
	// Variant support
	HasVariants      bool              `json:"has_variants" firestore:"has_variants"`
	Variants         []ProductVariant  `json:"variants,omitempty" firestore:"variants,omitempty"`
	AvailableColors  []string          `json:"available_colors,omitempty" firestore:"available_colors,omitempty"`
	AvailableSizes   []string          `json:"available_sizes,omitempty" firestore:"available_sizes,omitempty"`
}

// Removed individual structs as we're using map[string]interface{} 
// to handle the flexible data structure from Firestore

File: tripund_mobile/lib/models/product.dart
class ProductVariant {
  final String id;
  final String color;
  final String size;
  final double price;
  final double? salePrice;
  final String sku;
  final int stockQuantity;
  final List<String>? images;
  final bool available;

  ProductVariant({
    required this.id,
    required this.color,
    required this.size,
    required this.price,
    this.salePrice,
    required this.sku,
    required this.stockQuantity,
    this.images,
    required this.available,
  });

  factory ProductVariant.fromJson(Map<String, dynamic> json) {
    double parsePrice(dynamic value) {
      if (value == null) return 0;
      if (value is int) return value.toDouble();
      if (value is double) return value;
      if (value is String) return double.tryParse(value) ?? 0;
      return 0;
    }

    return ProductVariant(
      id: json['id'] ?? '',
      color: json['color'] ?? '',
      size: json['size'] ?? '',
      price: parsePrice(json['price']),
      salePrice: json['sale_price'] != null ? parsePrice(json['sale_price']) : null,
      sku: json['sku'] ?? '',
      stockQuantity: json['stock_quantity'] ?? 0,
      images: json['images'] != null ? List<String>.from(json['images']) : null,
      available: json['available'] ?? true,
    );
  }

  double get displayPrice => salePrice ?? price;
}

class Product {
  final String id;
  final String name;
  final String description;
  final String? shortDescription;
  final double price;
  final double? salePrice;
  final List<String> images;
  final List<String> categories;
  final List<String>? tags;
  final String sku;
  final int stock;
  final Map<String, dynamic>? attributes;
  final double? rating;
  final int? reviewCount;
  final bool hasVariants;
  final List<ProductVariant>? variants;
  final List<String>? availableColors;
  final List<String>? availableSizes;

  Product({
    required this.id,
    required this.name,
    required this.description,
    this.shortDescription,
    required this.price,
    this.salePrice,
    required this.images,
    required this.categories,
    this.tags,
    required this.sku,
    required this.stock,
    this.attributes,
    this.rating,
    this.reviewCount,
    this.hasVariants = false,
    this.variants,
    this.availableColors,
    this.availableSizes,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    // Handle price - API returns price in rupees already
    double parsePrice(dynamic value) {
      if (value == null) return 0;
      if (value is int) return value.toDouble();
      if (value is double) return value;
      if (value is String) return double.tryParse(value) ?? 0;
      return 0;
    }
    
    // Parse variants if present
    List<ProductVariant>? variantList;
    if (json['variants'] != null && json['variants'] is List) {
      variantList = (json['variants'] as List)
          .map((v) => ProductVariant.fromJson(v))
          .toList();
    }

    // Parse available colors and sizes
    List<String>? colors;
    if (json['available_colors'] != null) {
      colors = List<String>.from(json['available_colors']);
    }
    List<String>? sizes;
    if (json['available_sizes'] != null) {
      sizes = List<String>.from(json['available_sizes']);
    }

    return Product(
      id: json['id'] ?? '',
      name: json['name'] ?? json['title'] ?? '',
      description: json['description'] ?? '',
      shortDescription: json['short_description'],
      price: parsePrice(json['price']),
      salePrice: json['sale_price'] != null ? parsePrice(json['sale_price']) : null,
      images: List<String>.from(json['images'] ?? []),
      categories: List<String>.from(json['categories'] ?? []),
      tags: json['tags'] != null ? List<String>.from(json['tags']) : null,
      sku: json['sku'] ?? '',
      stock: json['stock_quantity'] ?? json['stock'] ?? 0,
      attributes: json['attributes'] is List ? null : json['attributes'],
      rating: json['rating']?.toDouble(),
      reviewCount: json['review_count'],
      hasVariants: json['has_variants'] ?? false,
      variants: variantList,
      availableColors: colors,
      availableSizes: sizes,
    );
  }

  double get displayPrice => salePrice ?? price;
  
  double get discountPercentage {
    if (salePrice != null && salePrice! < price) {
      return ((price - salePrice!) / price * 100);
    }
    return 0;
  }

  bool get hasDiscount => salePrice != null && salePrice! < price;
  
  // Get cheapest variant price for display
  double get lowestPrice {
    if (hasVariants && variants != null && variants!.isNotEmpty) {
      double lowest = variants!.first.displayPrice;
      for (var variant in variants!) {
        if (variant.displayPrice < lowest) {
          lowest = variant.displayPrice;
        }
      }
      return lowest;
    }
    return displayPrice;
  }

  // Get variant by color and size
  ProductVariant? getVariant(String? color, String? size) {
    if (!hasVariants || variants == null) return null;
    return variants!.firstWhere(
      (v) => v.color == color && v.size == size,
      orElse: () => variants!.first,
    );
  }
}